### PR TITLE
[pytorch][embeddingbag] Parallelize the EmbeddingBag operator

### PR DESCRIPTION
**Pull request by jianyuh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32548 [pytorch][refactor] Explicitly use auto* for pointers
* **#27477 [pytorch][embeddingbag] Parallelize the EmbeddingBag operator**

We would like to add the intra-op parallelization support for the EmbeddingBag operator.

This should bring speedup for the DLRM benchmark:
https://github.com/pytorch/pytorch/pull/24385

Benchmark code:
```
rom __future__ import absolute_import, division, print_function, unicode_literals

import torch
import time

eb = torch.nn.EmbeddingBag(1000000, 64, mode='sum')

input = torch.LongTensor(1500).random_(0, 1000000)
offsets = torch.zeros(64, dtype=torch.int64)

niter = 10000
s = time.time()
for _ in range(niter):
    out = eb(input, offsets)
time_per_iter = (time.time() - s) / niter
print('time_per_iter', time_per_iter)
print('GB/s', (input.numel() * 64 * 4 + out.numel() * 4) / time_per_iter / 1e9)
```

The following results are single core on Skylake T6:
- Before our change (with the original caffe2::EmbeddingLookup)
time_per_iter 6.313693523406982e-05
GB/s 6.341517821789133

- After our change using the EmbeddingLookupIdx API which takes the offsets instead of lengths.
time_per_iter 5.7627105712890626e-05
GB/s 6.947841559053659

- With Intel's PR: https://github.com/pytorch/pytorch/pull/24385
time_per_iter 7.393271923065185e-05
GB/s 5.415518381664018

For multi-core performance, because Clang doesn't work with OMP, I can only see the single-core performance on SKL T6.

Differential Revision: [D17768404](https://our.internmc.facebook.com/intern/diff/D17768404/)

### CODE REVIEW

**Comment by jianyuh** [comment]
Update the code with the proposal from https://github.com/pytorch/pytorch/issues/29019

**Comment by yinghai** [comment]
Can we need up the tests?

**Comment by jianyuh** [comment]
- Updated the EmbeddingBag interface following https://github.com/pytorch/pytorch/issues/29019.
- Fixed the issues for the parallel results.
- Added the unit tests for the new interface.

**Comment by JianpingChen066** [comment]
Hi, Jianyu
I tried to merge the code here to run DLRM benchmark. Since parallel path is under new_offsets as true, so I modified EmbeddingBag the as below
    EE = nn.EmbeddingBag(n, m, mode="sum", sparse=True, **new_offsets=True**)

While I meet the error like below:
```
Traceback (most recent call last):
  File "dlrm_s_pytorch.py", line 802, in <module>
    Z = dlrm_wrap(X, lS_o, lS_i, use_gpu, device)
  File "dlrm_s_pytorch.py", line 727, in dlrm_wrap
    return dlrm(X, lS_o, lS_i)
  File "/home/jchen13/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "dlrm_s_pytorch.py", line 280, in forward
    return self.sequential_forward(dense_x, lS_o, lS_i)
  File "dlrm_s_pytorch.py", line 297, in sequential_forward
    z = self.interact_features(x, ly)
  File "dlrm_s_pytorch.py", line 248, in interact_features
    T = torch.cat([x] + ly, dim=1).view((batch_size, -1, d))
RuntimeError: invalid argument 0: Sizes of tensors must match except in dimension 1. Got 2048 and 2047 in dimension 0 at ../aten/src/TH/generic/THTensor.cpp:612
Min time per iteration = 999999
```

Any ideal for this ? Thanks

**Comment by jianyuh** [comment]
Hi @JianpingChen066 ,

If you want to use `new_offsets=True`, you need to add one more element in the end of offsets. Basically offsets will have num_bag + 1 element, and the last element is equivalent to the size of `indices` (Check #29019). The off-by-one error in ` Sizes of tensors must match except in dimension 1. Got 2048 and 2047 in dimension 0 at ../aten/src/TH/generic/THTensor.cpp:612` seems to be related to this.

**Comment by JianpingChen066** [comment]
Thanks Jianyu
After modify the modify the dlrm_s_pytorch.py as below :
```
--- a/dlrm_s_pytorch.py
+++ b/dlrm_s_pytorch.py
@@ -142,7 +142,8 @@ class DLRM_Net(nn.Module):
                 EE = QREmbeddingBag(n, m, self.qr_collisions,
                     operation=self.qr_operation, mode="sum", sparse=True)
             else:
-                EE = nn.EmbeddingBag(n, m, mode="sum", sparse=True)
+                EE = nn.EmbeddingBag(n, m, mode="sum", sparse=True, new_offsets=True)
+                # EE = nn.EmbeddingBag(n, m, mode="sum", sparse=True)

                 # initialize embeddings
                 # nn.init.uniform_(EE.weight, a=-np.sqrt(1 / n), b=np.sqrt(1 / n))
@@ -228,6 +229,8 @@ class DLRM_Net(nn.Module):
         ly = []
         for k, sparse_index_group_batch in enumerate(lS_i):
             sparse_offset_group_batch = lS_o[k]
+            new_offset = torch.tensor([len(sparse_index_group_batch)])
+            sparse_offset_group_batch = torch.cat((sparse_offset_group_batch, new_offset), 0)

             # embedding lookup
             # We are using EmbeddingBag, which implicitly uses sum operator.
```
The new EmbeddingBag API works well and can get the expected performance

By the way, for new_offsets, whether this parameter also need to be added for EmbeddingBag backward ?

Thanks

**Comment by gchanan** [comment]
Your benchmark doesn't have a comparison with the `new_offsets` case?  Why?

File: aten/src/ATen/native/cuda/EmbeddingBag.cu [hunk 1]
```diff
@@ -264,6 +265,13 @@ _embedding_bag_cuda(const Tensor &weight, const Tensor &indices,
 
   int64_t numIndices = indices.size(0);
   int64_t numBags = offsets.size(0);
+  if (new_offsets) {
+    // Check https://github.com/pytorch/pytorch/issues/29019
+    // We plan to add one more element in offsets, which is equal to the size of
+    // indices. Currently for cuda devices, we still use the legacy
+    // implementation even this flag is enabled.
+    numBags -= 1;
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/EmbeddingBag.cu]: what happens if the size is 0?

**Comment by gchanan** [comment]
new_offsets is not a good name.  Consider if you find an even better offset structure in the future.  Do you call it new_new_offsets?  The name should be descriptive of the format (or making it a kwarg-only parameter with a different name).

**Comment by gchanan** [comment]
The benchmarks I would have expected:
1) current algorithm (you have this)
2) new algorithm without new_offsets (you have this, I believe)
3) new algorithm new new_offsets
4) "new" algorithm, but doing the special casing of the last offset.

**Comment by dzhulgakov** [comment]
Agree on the naming concern. `offsets_with_last_index`, `include_last_offset`, `offset_format='with_last'` - but I don't have the perfect name

Also - can you use / extend benchmarks in https://github.com/pytorch/pytorch/blob/master/benchmarks/operator_benchmark/pt/embeddingbag_test.py instead?

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 2]
```diff
@@ -74,26 +75,51 @@ void index_select_add<float>(const Tensor &select_indices,
...
-      /*indices=*/select_indices_data,
-      /*offsets=*/offsets.data_ptr<int64_t>(),
-      /*weights=*/nullptr,
-      /*scale_bias=*/nullptr,
-      /*normalize_by_lengths=*/false,
-      /*out=*/output_data
-    );
+    if (new_offsets) {
+      auto output_size = offsets.numel() - 1;
+      at::parallel_for(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/EmbeddingBag.cpp]: why do you parallelize in this branch but not in another?

File: aten/src/ATen/native/cuda/EmbeddingBag.cu (hunk 1 above)
Comment (dzhulgakov) [inline aten/src/ATen/native/cuda/EmbeddingBag.cu]: you want to add error checking here and in CPU branch that the offsets aren't empty when the new style is used

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 8a956858:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8a9568581c25f991f6e8a8d34d1f74607ccad6e1).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4314093) caffe2_onnx_py2_gcc5_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89265276) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89265276))
<details>
<summary>
<code>Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_embedding_bags FAILED   [  2%]</code>
</summary>

```
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_conv_variable_length PASSED [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_convtranspose PASSED    [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_cos PASSED              [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_cumsum PASSED           [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_det SKIPPED             [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_dict PASSED             [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_dict_str PASSED         [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_dim PASSED              [  1%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_dropout PASSED          [  2%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_elu PASSED              [  2%] 
Jan 22 07:08:06 test/onnx/test_operators.py::TestOperators::test_embedding_bags FAILED   [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_empty_like PASSED       [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_empty_like_opset7 PASSED [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_equal PASSED            [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_erf PASSED              [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_exp PASSED              [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_expand PASSED           [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_flatten PASSED          [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_flatten2D PASSED        [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_fmod PASSED             [  2%] 
Jan 22 07:08:07 test/onnx/test_operators.py::TestOperators::test_frobenius_norm PASSED   [  2%] 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 3]
```diff
@@ -74,26 +75,51 @@ void index_select_add<float>(const Tensor &select_indices,
...
+          0, output_size, 1, [&](int64_t start_idx, int64_t end_idx) {
+            caffe2::EmbeddingLookupIdx(
+                /*block_size=*/ddim,
+                /*output_size=*/end_idx - start_idx,
+                /*index_size=*/select_indices.numel(),
+                /*data_size=*/src.size(0),
+                /*input=*/src_data,
+                /*indices=*/select_indices_data,
```
Comment (jspark1105) [inline aten/src/ATen/native/EmbeddingBag.cpp]: We can pass select_indices_data + offsets.data_ptr<int64_t>()[start_idx] and, for index_size, offsets.data_ptr<int64_t>()[end_idx] - offsets.data_ptr<int64_t>()[start_idx] . This should enable error handling (see comment below).

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 4]
```diff
@@ -197,8 +253,12 @@ void index_select_scale_add<float>(const Tensor &select_indices,
...
 }  // namespace
 
-static at::Tensor make_bag_size(const Tensor &offsets, const Tensor &indices,
-                                const int64_t mode, const bool requires_grad) {
+static at::Tensor make_bag_size(
+    const Tensor& offsets,
+    const Tensor& indices,
+    const int64_t mode,
+    const bool requires_grad,
+    bool include_last_offset) {
```
Comment (jspark1105) [inline aten/src/ATen/native/EmbeddingBag.cpp]: Do we need this argument here?

File: caffe2/perfkernels/embedding_lookup_idx.cc [hunk 5]
```diff
@@ -28,14 +28,14 @@ static bool EmbeddingLookupGenericSlowIdx(
     const float* scale_bias, // optional scale & bias params for uint8 input
     bool normalize_by_lengths,
     OutType* out) {
-  int64_t current = 0;
+  int64_t current = offsets[0];
```
Comment (jspark1105) [inline caffe2/perfkernels/embedding_lookup_idx.cc]: With the suggested changes on the caller side we can just set current to 0 here and at the end we can compare current == index_size

File: caffe2/perfkernels/embedding_lookup_idx.cc [hunk 6]
```diff
@@ -185,10 +185,8 @@ static bool EmbeddingLookupGenericSlowIdx(
       return;                                                                                         \
     }                                                                                                 \
     int64_t current = 0;                                                                              \
```
Comment (jspark1105) [inline caffe2/perfkernels/embedding_lookup_idx.cc]: Without the suggested changes on the caller side, this doesn't match with what EmbeddingLookupGenericSlowIdx is doing

File: caffe2/perfkernels/embedding_lookup_idx.cc [hunk 7]
```diff
@@ -185,10 +185,8 @@ static bool EmbeddingLookupGenericSlowIdx(
...
-    for (int m = 0; m < output_size; ++m) {                                                           \
-      for (int64_t i = offsets[m];                                                                    \
-           i < (m == output_size - 1 ? index_size : offsets[m + 1]);                                  \
-           ++i) {                                                                                     \
+    for (int m = 0; m < output_size - 1; ++m) {                                                       \
```
Comment (jspark1105) [inline caffe2/perfkernels/embedding_lookup_idx.cc]: Why output_size - 1 ? We still want to check OOB indices for m == output_size - 1

File: caffe2/perfkernels/embedding_lookup_idx.cc [hunk 8]
```diff
@@ -202,11 +200,6 @@ static bool EmbeddingLookupGenericSlowIdx(
         ++current;                                                                                    \
       }                                                                                               \
     }                                                                                                 \
-    CAFFE_ENFORCE_EQ(                                                                                 \
-        current,                                                                                      \
-        index_size,                                                                                   \
-        "Your input seems to be incorrect: the sum of lengths values should be "                      \
-        "the size of the indices tensor, but it appears not.");                                       \
   }
```
Comment (jspark1105) [inline caffe2/perfkernels/embedding_lookup_idx.cc]: With the suggested changes, we can compare current and index_size

File: aten/src/ATen/native/EmbeddingBag.cpp (hunk 2 above)
Comment (yinghai) [reply aten/src/ATen/native/EmbeddingBag.cpp]: I think it's still doable just that it's not as plain.

**Review by yinghai — COMMENTED** [review]
LGTM. Could you check the CI failures?

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 9]
```diff
@@ -74,26 +75,46 @@ void index_select_add<float>(const Tensor &select_indices,
...
-      /*input=*/src_data,
-      /*indices=*/select_indices_data,
-      /*offsets=*/offsets.data_ptr<int64_t>(),
-      /*weights=*/nullptr,
-      /*scale_bias=*/nullptr,
-      /*normalize_by_lengths=*/false,
-      /*out=*/output_data
-    );
+    int64_t output_size = offsets.numel() - 1;
+    auto offsets_data = offsets.data_ptr<int64_t>();
```
Comment (yinghai) [inline aten/src/ATen/native/EmbeddingBag.cpp]: If this is a pointer, can we explicitly use auto* ?

**Review by yinghai — APPROVED** [review]
Thanks.

File: aten/src/ATen/native/EmbeddingBag.cpp (hunk 9 above)
Comment (jianyuh) [reply aten/src/ATen/native/EmbeddingBag.cpp]: Addressing this in https://github.com/pytorch/pytorch/pull/32548

**Comment by jianyuh** [comment]
Regarding the unit test, we tests the current algorithm (without “include_last_offset”) and new algorithm (with “include_last_offsets”) with their corresponding reference implementations. The comparison with the new “include_last_offset” is covered in the unit tests.

Regarding use / extend benchmarks in https://github.com/pytorch/pytorch/blob/master/benchmarks/operator_benchmark/pt/embeddingbag_test.py, will post a new PR as a follow-up.

File: torch/nn/modules/sparse.py [hunk 10]
```diff
@@ -206,6 +206,9 @@ class EmbeddingBag(Module):
         sparse (bool, optional): if ``True``, gradient w.r.t. :attr:`weight` matrix will be a sparse tensor. See
                                  Notes for more details regarding sparse gradients. Note: this option is not
                                  supported when ``mode="max"``.
+        include_last_offset (bool, optional): if ``True``, :attr:`offsets` has one additional element, where the last element
+                                      is equivalent to the size of `indices`. This matches the CSR format. Note:
+                                      this option is currently only supported when ``mode="sum"``.
```
Comment (dzhulgakov) [inline torch/nn/modules/sparse.py]: we should support it for other options as well (even if it's not efficient yet). Can we add implementation to slice off the last element for other modes for now?

**Pull request closed without merging**